lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs

```diff
@@ -1407,3 +1431,66 @@ blockchain =
     ]
   where
     slot e s = SlotNo $ flatSlot (EpochLength 21600) (SlotId e s)
+
+prop_applyTxToUTxO_balance :: Property
+prop_applyTxToUTxO_balance =
+    forAllShrink genTx shrinkTx $ \tx ->
+    forAllShrink genUTxO shrinkUTxO $ \u ->
+        balance (applyTxToUTxO tx u)
+        === balance u
+          `TokenBundle.add`
+              balance (utxoFromTx tx)
+          `TokenBundle.difference`
+              balance (u `UTxO.restrictedBy` Set.fromList (inputs tx))
+ 
+prop_applyTxToUTxO_entries :: Property
+prop_applyTxToUTxO_entries =
+    forAllShrink genTx shrinkTx $ \tx ->
+    forAllShrink genUTxO shrinkUTxO $ \u ->
+        unUTxO (applyTxToUTxO tx u)
+        === unUTxO u
+          `Map.union`
+              unUTxO (utxoFromTx tx)
+          `Map.difference`
+              unUTxO (u `UTxO.restrictedBy` Set.fromList (inputs tx))
+    
+prop_filterByAddress_allOurs :: Property
+prop_filterByAddress_allOurs =
+    forAllShrink genUTxO shrinkUTxO $ \u ->
+        filterByAddress (const True) u === u
+
+prop_filterByAddress_noneOurs :: Property
+prop_filterByAddress_noneOurs =
+    forAllShrink genUTxO shrinkUTxO $ \u ->
+        filterByAddress (const False) u === mempty
+
+prop_filterByAddress_empty :: Bool -> Property
+prop_filterByAddress_empty b =
+    let
+        f = const b
+    in
+        filterByAddress f mempty === mempty
+
+prop_filterByAddress_balance_applyTxToUTxO :: Bool -> Property
+prop_filterByAddress_balance_applyTxToUTxO b = 
+    let
+        f = const b
+    in
+        forAllShrink genTx shrinkTx $ \tx ->
+            balance (filterByAddress f (applyTxToUTxO tx mempty))
+            === foldMap (\o -> do
+                    if f (address o) then tokens o else mempty
+                ) (outputs tx)
+
+prop_filterByAddress_filterByAddressM :: Bool -> Property
+prop_filterByAddress_filterByAddressM b =
+    let
+        f = const b
+    in
+        forAllShrink genUTxO shrinkUTxO $ \u ->
+            filterByAddress f u === runIdentity (filterByAddressM (pure . f) u)
```

Comment:
Is it possible to add a property that uses a real filtering function?

There's already a convenient helper function available we can use here: `addressParity`.

You could choose either of the following:
```hs
(== Even) . addressParity
(==  Odd) . addressParity
```
Either of the above functions will select around half of all addresses.

Then you could verify that all addresses in the resulting UTxO set really do match what you expect.

Reply:
Addressed in https://github.com/input-output-hk/cardano-wallet/pull/2860/commits/d275095bc273a15bf21f117564e8ef80adc8d0bb (#2860).